### PR TITLE
Kifi Libraries Basics (Objects + Database + Tests & stuff)

### PR DESCRIPTION
**Pull request by ahsu1230**
@eishay @lawzlo @2is10 @martinraison @andrewconner @leogrim @stkem

### CODE REVIEW

**Comment by andrewconner** [comment]
Wahoo! Will review.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+object Library {
+  implicit val format = (
+    (__ \ 'id).formatNullable(Id.format[Library]) and
+      (__ \ 'createdAt).format(DateTimeJsonFormat) and
+      (__ \ 'updatedAt).format(DateTimeJsonFormat) and
+      (__ \ 'externalId).format(ExternalId.format[Library]) and
+      (__ \ 'name).format[String] and
+      (__ \ 'ownerId).format[Id[User]] and
+      (__ \ 'privacy).format[LibraryPrivacy] and
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: might consider "library visibility" instead of "library privacy"

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+case class Library (
+                 id: Option[Id[Library]] = None,
+                 createdAt: DateTime = currentDateTime,
+                 updatedAt: DateTime = currentDateTime,
+                 externalId: ExternalId[Library] = ExternalId(),
+                 name: String,
+                 ownerId: Id[User],
+                 privacy: LibraryPrivacy = LibraryPrivacy.PUBLIC,
+                 description: Option[String] = None,
+                 // Token name code: Set[String]
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: what is that?

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                 seq: SequenceNumber[Library] = SequenceNumber.ZERO
+                 ) extends ModelWithExternalId[Library] with ModelWithState[Library] with ModelWithSeqNumber[Library] {
+
+  def withId(id: Id[Library]) = this.copy(id = Some(id))
+  def withUpdateTime(now: DateTime) = this.copy(updatedAt = now)
+  def withName(myName: String) = this.copy(name = myName)
+  def withOwner(id: Id[User]) = this.copy(ownerId = id)
+  def withExternalId(id: ExternalId[Library]) = copy(externalId = id)
+  def withPrivacy(level: String) = this.copy(privacy = level match {
+    case "private" => LibraryPrivacy.PRIVATE
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: more safe would be to use the `LibraryPrivacy`value as in 

``` scala
case LibraryPrivacy.PRIVATE.VALUE => LibraryPrivacy.PRIVATE
```

I would also rather see this code in the `LibraryPrivacy` object. Maybe an Apply value

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    case "shared" => LibraryPrivacy.SHARED
+    case "public" => LibraryPrivacy.PUBLIC
+  })
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]:
``` scala
case unknown => throw new Exception(s"unknown LibraryPrivacy name $LibraryPrivacy")
```

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  extends JsonCacheImpl[LibraryExternalIdKey, Library](stats, accessLog, innermostPluginSettings, innerToOuterPluginSettings:_*)
+
+
+object LibraryStates extends States[Library] {
+  val PENDING = State[Library]("pending")
+  val BLOCKED = State[Library]("blocked")
+  val INCOMPLETE_SIGNUP = State[Library]("incomplete_signup")
+}
+
+case class LibraryPrivacy(value: String)
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Usually I would use a `sealed trait` to imitate java enums as in 

``` scala
object Test {
  sealed trait Parent
  case object A extends Parent
  case object B extends Parent
  case object C extends Parent
}
```

This way you can't create more `Parent` object anywhere around the code and the compiler will let you know if you missed one in the iteration.

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    libraryId bigint(20) NOT NULL,
+    userId bigint(20) NOT NULL,
+    permission varchar(20) NOT NULL,
+    state varchar(20) NOT NULL,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+
+    PRIMARY KEY (id),
+    FOREIGN KEY (userId) REFERENCES user(id),
```
Comment (eishay) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: always explicitly name keys (except primary key) so it will be easy to read their name when exceptions happes.
when naming prefix the name with <table name>_f_<column>
use `q` instead of `f` if its a unique constraint
for example

``` sql
CONSTRAINT `library_member_f_user` FOREIGN KEY (`user_id`) REFERENCES `user` (`id`)
```

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    privacy varchar(20),
+    tokens varchar(512) NOT NULL,
+
+    PRIMARY KEY (id),
+    FOREIGN KEY (ownerId) REFERENCES user(id)
+);
+
+CREATE TABLE library_member (
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    libraryId bigint(20) NOT NULL,
```
Comment (eishay) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: use underscores instead of camel case `library_id`

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import com.google.inject.{Inject, Singleton, ImplementedBy}
+import com.keepit.common.db.{ExternalId, Id}
+import com.keepit.common.db.slick._
+import com.keepit.common.logging.Logging
+import com.keepit.common.time.Clock
+
+@ImplementedBy(classOf[LibraryMemberRepoImpl])
+trait LibraryMemberRepo extends Repo[LibraryMember] with RepoWithDelete[LibraryMember] with SeqNumberFunction[LibraryMember]{
+
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: no need for 

```
{

}
```

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+}
+
+
+@Singleton
+class LibraryMemberRepoImpl @Inject()(
+                                 val db: DataBaseComponent,
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: why use `val` can't they be a private member of the class?

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+object LibraryMemberStates extends States[LibraryMember]
+
+case class LibraryMemberPrivacy(value: String)
+object LibraryMemberPrivacy {
+  implicit def format[T]: Format[LibraryMemberPrivacy] =
+    Format(__.read[String].map(LibraryMemberPrivacy(_)), new Writes[LibraryMemberPrivacy]{ def writes(o: LibraryMemberPrivacy) = JsString(o.value) })
+  val NO_ACCESS = LibraryMemberPrivacy("no_access")
+  val READ_ONLY = LibraryMemberPrivacy("read_only")
+  val READ_WRITE = LibraryMemberPrivacy("read_write")
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: We might also consider the name `LibraryAccess` or `LibraryAccessLevel`.

`NO_ACCESS` might be unnecessary, redundant with the `INACTIVE` state.

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+  override def toString: String = s"LibraryMember[id=$id,libraryId=$libraryId,userId=$userId,permission=$permission,state=$state]"
+}
+
+object LibraryMember {
+  implicit def format = (
+  (__ \ 'id).formatNullable(Id.format[LibraryMember]) and
+    (__ \ 'libraryId).format[Id[Library]] and
+    (__ \ 'userId).format[Id[User]] and
+    (__ \ 'permission).format[LibraryMemberPrivacy] and
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: I’d suggest having the field/column name and type name match. Maybe `access` or `access_level`. `permissions` could work, but that often connotes a bit-mask kind of scheme.

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+import com.keepit.common.db._
+import com.keepit.common.logging.AccessLog
+import com.keepit.common.time._
+import org.joda.time.DateTime
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+
+import scala.concurrent.duration.Duration
+
+case class LibraryMember(
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: What does everyone think of `LibraryMembership`? I think it would make it clearer that this entity represents a relationship (between a `User` and a `Library`) and is not an alternate representation of a person (akin to `User`).

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  def toKey(): String = externalId.id
+}
+class LibraryExternalIdCache(stats: CacheStatistics, accessLog: AccessLog, innermostPluginSettings: (FortyTwoCachePlugin, Duration), innerToOuterPluginSettings: (FortyTwoCachePlugin, Duration)*)
+  extends JsonCacheImpl[LibraryExternalIdKey, Library](stats, accessLog, innermostPluginSettings, innerToOuterPluginSettings:_*)
+
+
+object LibraryStates extends States[Library] {
+  val PENDING = State[Library]("pending")
+  val BLOCKED = State[Library]("blocked")
+  val INCOMPLETE_SIGNUP = State[Library]("incomplete_signup")
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I think custom states deserve comments. If possible, I think it would be great to minimize the number of states here. We may not need to explicitly represent an incomplete user signup. That’s probably already represented in the `User` entity. Maybe that could be collapsed into `PENDING` here? Or maybe we shouldn’t create libraries for users until they’ve completed signup?

**Comment by 2is10** [comment]
Nice work, by the way. Not trying to be critical, just trying to help us arrive at a domain model that we won’t immediately regret, as we have in the past. Database naming is quite hard to change.

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  def table(tag: Tag) = new LibraryTable(tag)
+  initTable()
+
+  private def getLibrary(id: Column[Id[Library]]) =
+    for(f <- rows if f.id is id) yield f
+  private val getCompiled = Compiled(getLibrary _)
+
+  override def get(id: Id[Library])(implicit session: RSession): Library = {
+    idCache.getOrElse(LibraryIdKey(id)) {
+      getCompiled(id).first
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala:50]: nice, a bit of boilerplate :-(

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  def table(tag: Tag) = new LibraryMemberTable(tag)
+
+  initTable()
+
+
+  private def getLibraryMember(id: Column[Id[LibraryMember]]) =
+    for(f <- rows if f.id is id) yield f
+  private val getCompiled = Compiled(getLibraryMember _)
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: Maybe we can hide `getLibraryMember` more?

``` scala
private val getCompiled = {
  def getLibraryMember(id: Column[Id[LibraryMember]]) = for(f <- rows if f.id is id) yield f
  Compiled(getLibraryMember _)
}
```

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: For the URL. `urlToken` may be more clear?

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 3 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: We actually can probably remove most of these. Not sure the origin of this pattern (could have been me?), but they do what `copy()` does, with more lines of code.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 5 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I think we moved away from this because it complicates the serialization and deserialization patterns. The other instances in the codebase of custom types like this are similar. Talk tomorrow?

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala (hunk 10 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: Oo, I like `LibraryAccess`. We talked about `NO_ACCESS` earlier — potentially could be used for explicit blacklisting. Since that's no where in the roadmap, Jared's probably right, we should keep it lean and add it later if we need it.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 13 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I think this was just copy/paste from User. We aren't tracking incomplete, pending, blocked, etc. Active/inactive is plenty.

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala (hunk 15 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: Yeah, or tuck it inside. I'm going to actually move the default `DbRepo` methods to use compiled queries, and we'll come up with the most readable + cleanest syntax.

Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: nice, these methods are by far the most used (i think) so it would be a large gain

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 3 above)
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I think that some of it was me. Maybe when there was a side effect or verification to be made.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 5 above)
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: sure

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 2 above)
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: or “[slug](http://en.wikipedia.org/wiki/Semantic_URL#Slug)”

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 1 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Good idea, I changed to use visibility. Might be changing the 'privacy' name in LibraryMember as well

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 5 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: is there a verdict for this yet? We could talk =)

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+case class LibraryPrivacy(value: String)
+object LibraryPrivacy {
+  implicit def format[T]: Format[LibraryPrivacy] =
+    Format(__.read[String].map(LibraryPrivacy(_)), new Writes[LibraryPrivacy]{ def writes(o: LibraryPrivacy) = JsString(o.value) })
+  val PUBLIC = LibraryPrivacy("public")
+  val PRIVATE = LibraryPrivacy("private")
```
Comment (ymatsuda) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Is "private" different from a library shared by only its owner?

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala (hunk 9 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMemberRepo.scala]: I kind of copied the way other Repos like KeepRepo & UserRepo defined their classes. User/KeepRepo also use val (and not private val). Would LibraryMemberRepo be different in how it handles the db, clock, etc. members?

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala (hunk 10 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: Changed to LibraryAccess instead & removed NO_ACCESS for now.

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala (hunk 11 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: changed

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I like slug.

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala (hunk 12 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: I like.

Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: I like that. =)

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 13 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Ah woops, totally missed that. Yea, it was just a copy/paste from User
For LibraryStates, I'll just keep it as ACTIVE vs. INACTIVE for now.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 16 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: @ahsu1230 Since library member level access levels are handled in `LibraryMembership` (after the rename), what if we have two options here: `LIMITED` and `PUBLICLY_VIEWABLE`?

**Comment by ahsu1230** [comment]
No problem @2is10! These are some great insights. =) Thanks

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 2 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: I like slug too =) but it's still commented out for now, so we can add the whole feature later!

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 4 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: fixed

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 16 above)
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: or just a boolean `public` or `publicly_viewable`?

Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: we've decided to keep 3 stages:
SECRET, LIMITED, PUBLICLY_VIEWABLE

for easy & quick denormalization for when libraries are specifically limited to only the owner

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql (hunk 6 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: changed!

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    external_id varchar(36) NOT NULL,
+    state varchar(20) NOT NULL,
+
+    name varchar(256) NOT NULL,
+    ownerId bigint(20) NOT NULL,
+    description varchar(512),
+    privacy varchar(20),
```
Comment (andrewconner) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: NOT NULL

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+# --- !Ups
+
+CREATE TABLE library (
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    external_id varchar(36) NOT NULL,
+    state varchar(20) NOT NULL,
+
+    name varchar(256) NOT NULL,
```
Comment (andrewconner) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: UNIQUE on this? @2is10 thoughts?

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala (hunk 12 above)
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMember.scala]: +1

File: kifi-backend/modules/common/conf/evolutions/shoebox/191.sql (hunk 18 above)
Comment (eishay) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: UNIQUE on `(name, ownerId)`.
Would the url be:
http://www.kifi.com/eishay/myLibrary
If so, what happens if I change ownership? Should we preserve the path?

Comment (2is10) [inline kifi-backend/modules/common/conf/evolutions/shoebox/191.sql]: unique (name, ownerId) seems to make sense, no?

**Comment by 2is10** [comment]
Might want to consider having a `LibraryName` String wrapper type that enforces some simple restrictions, like a maximum length, no control characters (including tabs, newlines) and no quotation marks and perhaps forbidding certain other punctuation. We can always loosen up restrictions later, but it’s harder to add or tighten restrictions later.

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import play.api.libs.functional.syntax._
+import play.api.libs.json._
+
+import scala.concurrent.duration.Duration
+
+case class LibraryMembership(
+                     id: Option[Id[LibraryMembership]] = None,
+                     libraryId: Id[Library],
+                     userId: Id[User],
+                     access: LibraryMembershipAccess = LibraryMembershipAccess.READ_ONLY,
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala]: I don't see why we should have a default value. Opens us up to bugs where we forget to specify one, and it works.

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+import scala.concurrent.duration.Duration
+
+case class Library (
+                 id: Option[Id[Library]] = None,
+                 createdAt: DateTime = currentDateTime,
+                 updatedAt: DateTime = currentDateTime,
+                 externalId: ExternalId[Library] = ExternalId(),
+                 name: String,
+                 ownerId: Id[User],
+                 visibility: LibraryVisibility = LibraryVisibility.PUBLICLY_VIEWED,
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: See other comment about default values.

File: kifi-backend/modules/common/conf/evolutions/shoebox/192.sql [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    external_id varchar(36) NOT NULL,
+    state varchar(20) NOT NULL,
+
+    name varchar(256) NOT NULL,
+    owner_id bigint(20) NOT NULL,
+    description varchar(512),
+    visibility varchar(20) NOT NULL,
+    tokens varchar(50) NOT NULL,
```
Comment (andrewconner) [inline kifi-backend/modules/common/conf/evolutions/shoebox/192.sql]: slug

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  import DBSession._
+  import db.Driver.simple._
+  private val sequence = db.getSequence[LibraryMembership]("library_member_sequence")
+
+  type RepoImpl = LibraryMemberTable
+
+  class LibraryMemberTable(tag: Tag) extends RepoTable[LibraryMembership](db, tag, "library_member") with SeqNumberColumn[LibraryMembership] {
+    def libraryId = column[Id[Library]]("library_id", O.NotNull)
+    def userId = column[Id[User]]("user_id", O.Nullable)
+    def permission = column[LibraryMembershipAccess]("privacy", O.NotNull)
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryMembershipRepo.scala]: `def access = column[LibraryMembershipAccess]("access", O.NotNull)`

File: kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+  import scala.slick.lifted.Query
+  import DBSession._
+  import db.Driver.simple._
+  private val sequence = db.getSequence[Library]("library_sequence")
+
+  type RepoImpl = LibraryTable
+  class LibraryTable(tag: Tag) extends RepoTable[Library](db, tag, "library") with ExternalIdColumn[Library] with SeqNumberColumn[Library] {
+    def name = column[String]("name", O.NotNull)
+    def ownerId = column[Id[User]]("userId", O.Nullable)
+    def privacy = column[LibraryVisibility]("privacy", O.NotNull)
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/LibraryRepo.scala]: `def visibility = column[LibraryVisibility]("visibility", O.NotNull)`

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+                 seq: SequenceNumber[Library] = SequenceNumber.ZERO
+                 ) extends ModelWithExternalId[Library] with ModelWithState[Library] with ModelWithSeqNumber[Library] {
+
+  def withId(id: Id[Library]) = this.copy(id = Some(id))
+  def withUpdateTime(now: DateTime) = this.copy(updatedAt = now)
+
+  def withPrivacy(level: String) = this.copy(visibility = level match {
+    case "private" => LibraryVisibility.SECRET
+    case "shared" => LibraryVisibility.LIMITED
+    case "public" => LibraryVisibility.PUBLICLY_VIEWED
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: `PUBLICLY_VIEWABLE` perhaps? with "publicly_viewable"

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+                 // slug: Set[String]
+                 state: State[Library] = LibraryStates.ACTIVE,
+                 seq: SequenceNumber[Library] = SequenceNumber.ZERO
+                 ) extends ModelWithExternalId[Library] with ModelWithState[Library] with ModelWithSeqNumber[Library] {
+
+  def withId(id: Id[Library]) = this.copy(id = Some(id))
+  def withUpdateTime(now: DateTime) = this.copy(updatedAt = now)
+
+  def withPrivacy(level: String) = this.copy(visibility = level match {
+    case "private" => LibraryVisibility.SECRET
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: "secret"

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    case "shared" => LibraryVisibility.LIMITED
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: "limited"

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+  extends JsonCacheImpl[LibraryExternalIdKey, Library](stats, accessLog, innermostPluginSettings, innerToOuterPluginSettings:_*)
+
+
+object LibraryStates extends States[Library]
+
+case class LibraryVisibility(value: String)
+object LibraryVisibility {
+  implicit def format[T]: Format[LibraryVisibility] =
+    Format(__.read[String].map(LibraryVisibility(_)), new Writes[LibraryVisibility]{ def writes(o: LibraryVisibility) = JsString(o.value) })
+  val PUBLICLY_VIEWED = LibraryVisibility("public")
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Update strings

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala [hunk 28]
```diff
@@ -0,0 +1,61 @@
...
+
+object LibraryMembershipStates extends States[LibraryMembership]
+
+case class LibraryMembershipAccess(value: String)
+object LibraryMembershipAccess {
+  implicit def format[T]: Format[LibraryMembershipAccess] =
+    Format(__.read[String].map(LibraryMembershipAccess(_)), new Writes[LibraryMembershipAccess]{ def writes(o: LibraryMembershipAccess) = JsString(o.value) })
+  val READ_ONLY = LibraryMembershipAccess("read_only")
+  val READ_WRITE = LibraryMembershipAccess("read_write")
+}
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala:59]: Might be nice to shorten this to `LibraryAccess`. It’s the library that’s accessed, after all.

Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala:59]: Do you think we’ll need to be able to distinguish the different kinds of writes (like keep insertions, keep edits, and keep removals)? Even if we don’t want to support more write granularity initially, it might pay to choose names now that will not be ambiguous in the event that we do want to add more granularity later. I guess `READ_INSERT` could work with the naming scheme above if we wanted to have an access level for reading and keeping only in the future. (`WRITE` would continue to signify all kinds of writes.)

File: kifi-backend/modules/common/app/com/keepit/model/Library.scala (hunk 24 above)
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: Isn’t the word "viewable" redundant with the type name `LibraryVisibility`? I think `PUBLIC` would be clear. Or `ANYONE`.

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Library.scala]: That works. `ANYONE` is pretty clear.

**Comment by andrewconner** [comment]
Looks good. Good job!

File: kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala (hunk 28 above)
Comment (ahsu1230) [inline kifi-backend/modules/common/app/com/keepit/model/LibraryMembership.scala:59]: I think that's a good idea from our meeting earlier today.
Do you think the owner may need a MANAGE level?
Would owner give other collaborators the MANAGE access level so other people can also give these access rights?
